node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/task/util/file/FileFinder.java

```diff
@@ -0,0 +1,251 @@
+    }
+
+    /** Applies a given consumer to all the matching {@link FileFinder.FileAttributes} */
+    public void forEach(Consumer<FileAttributes> action) {
+        applyForEachToMatching(basePath, matcher, maxDepth, action);
+    }
+    
+
+    /**
+     * <p> This method walks a file tree rooted at a given starting file. The file tree traversal is
+     * <em>depth-first</em>: The filter function is applied in pre-order (NLR), but the given
+     * {@link Consumer} will be called in post-order (LRN).
+     */
+    private void applyForEachToMatching(Path basePath, Predicate<FileAttributes> matcher,
+                                        int maxDepth, Consumer<FileAttributes> action) {
+        try {
+            // Only need to traverse as deep as we want to match, unless we want to match everything in directories
+            // already matched
+            Files.walkFileTree(basePath, Collections.emptySet(), maxDepth, new SimpleFileVisitor<Path>() {
+                private final Stack<FileAttributes> matchingDirectoryStack = new Stack<>();
+                private int currentLevel = -1;
+
+                @Override
+                public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
+                    currentLevel++;
+
+                    FileAttributes attributes = new FileAttributes(dir, attrs);
+                    if (currentLevel > 0 && matcher.test(attributes))
+                        matchingDirectoryStack.push(attributes);
+
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
+                    // When we find a directory at the max depth given to Files.walkFileTree, the directory
+                    // will be passed to visitFile() rather than (pre|post)VisitDirectory
+                    if (attrs.isDirectory()) {
+                        preVisitDirectory(file, attrs);
+                        return postVisitDirectory(file, null);
+                    }
+
+                    FileAttributes attributes = new FileAttributes(file, attrs);
+                    if (matcher.test(attributes))
+                        action.accept(attributes);
+
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult postVisitDirectory(Path dir, IOException exc) {
+                    if (!matchingDirectoryStack.isEmpty())
+                        action.accept(matchingDirectoryStack.pop());
+
+                    currentLevel--;
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+        } catch (NoSuchFileException ignored) {
+
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static int deleteRecursively(FileAttributes fileAttributes) {
```

Comment:
recursiveDelete(ContainerDataImpl.java) is much easier to understand and maintain. Move recursiveDelete to UnixPath and call from here?

Reply:
Added to UnixPath, removed from ContainerDataImpl